Switch from distutils to setuptools

vkuznet opened the pull request:
Fixes #12256 

#### Status
in development

#### Description
Provide switch from `distutils` to `setuptools` library for all scripts and tools used by WMCore code-base. This PR provides the following:
- `.github/workflows/build_and_install_locally.yaml` new CI/CD workflow to locally build and install WM packages
- `bin/test_local_build_and_install.sh` script which imitates CI/CD workflow and relies on virtual environment to build and install WM packages locally
- modified version of all setup scripts used in build process of WM packages. The current WM build system follows these steps:
  - `setup.py` script can be used to build wmcore package, e.g. `python setup.py sdist`
  - `setup_template.py` script is used by CI/CD workflows to build concrete WM package, e.g. reqmgr2
     - CI/CD workflow (e.g. `.github/workflows/pypi_build_and_images.yaml`) copies `setup_tempate.py` into `setup.py` and properly setup package name
     - it also creates package specific version of `requirements.txt`
     - finally, it runs `python3 setup.py clean sdist` to create distribution tar ball.

**NOTE 1:** the WM build steps run within CI/CD pipeline as a matrix and therefore overwritten `setup.py` and `requirements.txt` files never cause any problems, i.e. they are within separate docker images during CI/CD execution, i.e. within specific matrix build.

**NOTE 2:** 
- The provided `bin/test_local_build_and_install.sh` script  only build and install WM packages in local virtual environment. It does not verifies their functionality though as it requires additional setup like databases, htcondor, gfal libraries.

- It also install all WM packages within single virtual environment. Another approach would be use individual virtual environment areas for individual WM packages, and if it will be desired I can provide a new script or modify `bin/test_local_build_and_install.sh` accordingly. The issue with multiple virtual environments that it waste disk space as WM packages share dependencies among each other. Therefore, I opted with single virtual environment approach.

- Local builds and installation skips `htcondor` and `gfal` dependencies as they are very centric for CERN based Linux OS. All other dependencies are OS agnostic and presented in local build/install procedure which can be performed on any Linux, and macOS versions.

**NOTE 3:** I removed `setup_wmcore.py` from the repository in this [commit](https://github.com/dmwm/WMCore/pull/12259/commits/d178016107d573e9790f11cd4fc7a55a22aec6ab) since I was unable to find out usage of this code anywhere in WMCore repository.

#### Is it backward compatible (if not, which system it affects?)
Most likely YES, but I'm not 100% sure.

I performed local installation and compared tarballs using distutils and setuptools, but no actual tests were done with individual packages as it requires their upload to PyPi and installation elsewhere.

**FINAL NOTE:** after we merge this PR we need to do the following steps:
- create new release candidate and check new CI/CD pipeline
- create new issue to merge two CI/CD pipelines (one proposed here and the one with pushing to PyPI) to have local build as pre-requisite step before merging to PyPi

#### Related PRs

#### External dependencies / deployment changes

vkuznet committed:
switch from distutils to setuptools

vkuznet commented:
@amaltaro , @todor-ivanov , @mapellidario I asked you to review this PR and I want to let you know that I identified that we have `setup_wmcore.py` file in repository that is not used anywhere. I don't know if we need it and I'll be happy to remove it but before that I want to inform you about this finding as well along the changes I made in this PR.

amaltaro commented:
@vkuznet if you couldn't find any reference to it, my suggestion is to make a new commit deleting it and making an explicit note in the initial PR description that it is being deleted because it is unused.

vkuznet commented:
> @vkuznet if you couldn't find any reference to it, my suggestion is to make a new commit deleting it and making an explicit note in the initial PR description that it is being deleted because it is unused.

ok, removed and adjusted the description.

vkuznet commented:
@amaltaro , @mapellidario here is gentle ping for you to review this PR as it sits for two weeks already without any actions on your side.

amaltaro reviewed (commented):
Thank you for providing these changes, Valentin. Here are some questions that I have on it:
1) can you please elaborate on the new CI/CD workflow? Is it temporary, or you wanted to persist it in the repository? What is the benefit of it compared to the other workflows already in?
2) why did you skip the htcondor and gfal dependencies? I suspect we don't do this in the current setup, especially for htcondor.
3) I understand you compared the tarball package constructed from distutils vs setuptools, correct? Did you compare everything or only python-based files? Did you try to import any WMCore or Utils library?
4) In terms of validating these changes, I wonder if we can manually run a workflow either against pypi-test or production (but then using a tag that we failed to built in the past).

Do you have any other concerns and/or points that would be nice to check, either before or after merging these in?

mapellidario reviewed (approved):
I made sure that I could run the `bin/test_local_build_and_install.sh`,  so I approve this PR, I love the switch to setuptools which allows us to use a python version >= 3.12. Terrific work, thanks Valentin!

However, I do not like that we skip the installation of htcondor and gfal. 

[edit] i did not see the previous comment by valenting, i am changing my feedback

~~I suggest to just add htcondor, it does not give any problems, it worked for me on the python:3.8 docker image.~~ htcondor can be installed easily via pip from pypi on any linux docker image. not as nice as working osx, I know, we can skip it.

gfal it's a different story, and I have to admin that I never liked how we treat our dependency on gfal. I dislike installing it from [miniconda](https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/gfal/Dockerfile), the only pacakge that we take from it, and we are lucky it does not clash with anything else in our debian image. I understand that the rationale for our past decisions was getting something done, not having the most elegant solution. I understand that the purpose of this script is to make sure that we can install our packages and not that our packages work well alongside gfal2, I understand that gfal is used only by ms-unmerged.

Nonetheless, I want to provide and example of how we could deal with the gfal dependency so that we can script it if needed, without copying it from the miniconda docker image.

<details><summary>Install gfal2-python with pip on debian</summary>

I used the official `python:3.8` docker image, which is based on debian 12.7.

```bash
git clone https://github.com/vkuznet/wmcore
cd wmcore
git fetch origin
git checkout -t origin/disutils2setuptools
docker run \
        -v $PWD:/home/dmwm/wmcore/WMCore:Z \
        -it --entrypoint /bin/bash \
        python:3.8
```

then I installed gfal2 with

```bash
apt update
# apt upgrade
apt install cmake
apt install gfal2* libgfal2* python3-gfal2*

# get boost libraries code
cd /tmp
# get boost version from what is installed by debian12 https://packages.debian.org/bookworm/libboost-all-dev
# do not install boost from debian repos because it targets python3.11, we need py3.8 here
wget https://archives.boost.io/release/1.74.0/source/boost_1_74_0.zip
unzip -q boost_1_74_0.zip
cd boost_1_74_0
./bootstrap.sh --prefix=boost_output --with-python=python3.8
# compile boost libraries
## on my laptop: -j16 -> 2min. -j1 -> 10min
time ./b2 install
# install boost libraries
cp -r boost_output/lib/libboost_* /usr/lib/x86_64-linux-gnu/
cp -r boost_output/include/boost/ /usr/include/

pip install --upgrade pip
pip install gfal2-python~=1.11.0
```

then i could run the new script (that skips htcondor/gfal2)

```bash
cd /home/dmwm/wmcore/WMCore
bash -x bin/test_local_build_and_install.sh
```

but I could also install our packages at their fullest, for example

```bash
pkg=wmagent
sed "s/PACKAGE_TO_BUILD/${pkg}/" setup_template.py > setup.py
python3 setup.py clean sdist bdist_wheel
ls -lh dist/
pip install -r requirements.txt
pip install --no-index --find-links=dist/ ${pkg}
pip list --format=freeze | grep ${pkg}
```

i do not know how many cores the default github runners have, i do not know how close or far we are from the cpu limit for our runners, I know that compiling the boost libraries every time for a simple pip install test maybe a waste

</details>
